### PR TITLE
Delete stale EndpointSlices on restart

### LINKED ISSUE #1416: Service discovery returning IPs of stale endpoints
Related to https://issues.redhat.com/browse/ACM-8423. 

The LH `EndpointSlice` still remained after a service was deleted. It seems the service was deleted while the LH agent wasn't running. We need to delete stale `EndpointSlices` on restart.

### PR DESCRIPTION
**Pull request by tpantelis**
If a local LH `EndpointSlices` exists but the corresponding `Service` doesn't exist, delete the `EndpointSlice`.

Also, a headless `Service` can have multiple `EndpointSlices` so reconcile the LH `EndpointSlices` on startup to handle a K8s `EndpointSlice` deleted while the controller wasn't running.

Fixes https://github.com/submariner-io/lighthouse/issues/1416

### CODE REVIEW

**Review by aswinsuryan — APPROVED** [review]

**Review by vthapar — APPROVED** [review]

File: pkg/agent/controller/endpoint_slice.go [hunk 1]
```diff
@@ -116,8 +118,31 @@ func (c *EndpointSliceController) onLocalEndpointSlice(obj runtime.Object, _ int
...
+
+	// Check if the associated Service exists and, if not, delete the EndpointSlice. On restart, it's possible the Service could've been
+	// deleted.
+	if op == syncer.Create {
+		_, found, _ := c.serviceSyncer.GetResource(serviceName, endpointSlice.Namespace)
+		if !found {
+			logger.Infof("The service %q for EndpointSlice \"%s/%s\" does not exist - deleting it",
+				serviceName, endpointSlice.Namespace, endpointSlice.Name)
+
+			err := c.syncer.GetLocalFederator().Delete(ctx, endpointSlice)
```
Comment (sridhargaddam) [inline pkg/agent/controller/endpoint_slice.go:134]: I hope this deletes the EndpointSlice from the Broker as well?

Comment (sridhargaddam) [inline pkg/agent/controller/endpoint_slice.go:134]: I hope this deletes the EndpointSlice from the Broker as well?

File: pkg/agent/controller/reconciliation_test.go [hunk 2]
```diff
@@ -231,6 +239,85 @@ var _ = Describe("Reconciliation", func() {
...
+
+				By("Restarting controllers")
+
+				restoreBrokerResources()
+				test.CreateResource(t.cluster1.localServiceImportClient.Namespace(test.LocalNamespace), localServiceImport)
+				test.CreateResource(t.cluster1.localEndpointSliceClient, localEndpointSlice)
+				test.CreateResource(t.cluster1.localServiceExportClient, serviceExport)
+				t.cluster1.createService()
+
+				fmt.Println("*******EPS 1: ", resource.ToJSON(localEndpointSlice))
```
Comment (sridhargaddam) [inline pkg/agent/controller/reconciliation_test.go]:
```suggestion

```

**Review by sridhargaddam — CHANGES_REQUESTED** [review]

File: pkg/agent/controller/endpoint_slice.go (hunk 1 above)
Comment (tpantelis) [reply pkg/agent/controller/endpoint_slice.go:134]: yup - the admiral syncer does that.

File: pkg/agent/controller/reconciliation_test.go (hunk 2 above)
Comment (tpantelis) [reply pkg/agent/controller/reconciliation_test.go]: oops

**Review by sridhargaddam — APPROVED** [review]

**Pull request merged**